### DISCUSSION TITLE
Advice for Entity design decision

### DISCUSSION THREAD

**Discussion by zirkelc** [tywalch/electrodb#314] [Q&A] [answered]
Hi,

I'm seeking advice on designing one or more entities. Let me explain my use case.

Imagine a multi-tenant smartphone scenario:

- A User belongs to a Tenant.
- A App belongs to a Tenant.
- A Homescreen belongs to a User.
  - A Homescreen contans multiple Widgets with a specific order in which they are rendered
  - A Widget is a reference to an App with a name and a size (small, medium, large).
  - An App can exists independently of being used as a Widget.

At the moment, I have identified three entities: User, App, and Homescreen. The Widget is not a separate entity but a custom array field within the Homescreen entity. Here's a simplified representation: [Playground](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAYziEAwqlxUokgIJUq2AEaoqcACqbppAL7cMNmAARIhwRo6wTkEA3OJGUHxWYFQMTgikALzBmlCoGAD6KWlw+XwMknAxcQlJlphgWXxwLKrqWgAUomBKXd1gkrCIAFw9fX38GlSaI0F1GEH4vWP6mLhYCTMAjAtLfcQYhiYzKbgA1gzS0kG7PotjDA7OrngjirvdbmXqAJIwr+99KbeGZWbB8ADmOzGY3IaCw5D+yQw6Du0NuALmv3+0MBXjgIMcWAhUJxZDgcIRI0c6AB6LGdL6RPgAA8XqNodIAgArMI0bGk6Snfmk7g4BCIoKCkki+IyKBrNwjADaQU+DB+MCCAF0Ad0GTizsLSVwxRKztLSbLZAr8WAVZjNTqRT5aai9b0GcoiukwC6AJSxaqJBQXaQNMBNFpqSaaTrdN5jAbwBBG7oTLQzUMW7orDBrDbBbZuvYGYy21UMM5Zm7Fh6ElxuXCp5L8dVULHsnFA8ugomQ4sw8moeFwRHU0yugGhjsJ0ndglg-u6skU0dU5FwSc4soVZsfPELvvZ3013ZMuCspudvqcqA88JX2c4wV7vomxASqUD6FW+XsW0qmqGrasu+rQoa144u+4ogqcx5jL+NrKkE06OqBW59C6eqol6qQ+v6gbdmAADqWAwOCcAKFks5sORlEdr2EKxN0qEjIx4LMck7DpGxhJMb0awAF49pIDAIAgQRgAAPsEFQwFgqCSJJMlBAgTAUVULqiERpF0TQ4a2BgGAMJoAA8ukUVQAB8gbxMGYB0DYeBGOQ-DhpGrQxnG15JsMkFgOmUwzI5FS4C5JB8Me+yHOWJznJcx65vmfBbNc9K1o8WANmyT6Aq2GqvvOwTsfBK7DpSSIohhYwOoVB7FXxS4irC5VrpVE44mB3S0ZZj7LkVFhWDY9j1s8njeOZZG9VZ7QoXwmhBH6tKnmM56Xs2FSSE4qyvi+-kITY1r-shQHto635jNBn5waBF1EEK+19Ihx12rM+y-Nqd3dFdsFpaSWFosWAOeskeG+KIAaiEAA)

In this setup, querying a specific User's Homescreen requires just one query. The Widgets on the Homescreen contain the necessary information for rendering (name, order, and size), with the order determined by the array index.

However, I face a problem when deleting an App: I must also delete all Widgets referencing that App from all Homescreens. To optimize this access pattern, I've considered treating Widget as its own entity with a secondary index for TenantID and AppID. This would make it straightforward to query and remove all Widgets tied to a particular App upon its deletion.

However, by having the Widgets as a separate entity, I lose the array index to order on the Homescreen. I've though about two approaches:

1. Keep the Homescreen with a Widgets fields that is simply an array of WidgetIDs. The array index is the order of the Widget on the Homescreen. That means I must query the Homescreen and all Widgets and render the Widgets according to the array indexes. However, when I delete an App, I must delete the Widgets entities and remove the WidgetIDs from the Homescreen entity.
2. Merge Homescreen and Widget into one Entity and add an order field to this Entity. However, re-arranging the position of a Widget means I must update all other Widgets as well. Even worse, when I add an App to the Homescreen to create a new Widget, I must query all other Widgets first to determine the last order value and increment it. 

---
I hope it's not too difficult to understand what I want to achieve :)

**Comment by tywalch** [comment]
Does this sound like I understand the crux of the challenge?:

- Users use Apps as Widgets, but an App exists whether or not it is actually used as a Widget
- When an App is deleted, you need to ensure that you also clean up any references it had as a Widget
- Regardless of how things are modeled, it is critical that it allows the user to control the order in which Widgets are presented.

Is that the extent of your requirements, am I missing anything?

**Reply by zirkelc** [reply]
Yes, that's right!

**Reply by zirkelc** [reply]
Not to bug you, but have you had a chance to think about it?

I'm not satisfied with the alternative solutions I came up with yet. I am currently leaning towards handling the deletion of an App as a DynamoDB Stream. The stream handler would have to query all Homescreen for a Tenant and see if the App was used as a Widget. I could add `appIds` as a top-level list attribute and use a query with `appIds contains($appId)` to push the filtering on the server-side.

**Reply by tywalch** [reply]
No it's ok, here's some general thoughts:

In an ideal world you're able to leverage an index to find all relevant app usages. In this case that could be that you store app usage, separate from the app order. That could be as an individual record per usage, like a reference record between app and user that's presence signals usage. 

In cases where you do have denormalize data, or write twice, determine: the actions/scenarios that interact with this data and rank them frequency they are encountered in real life. Their frequency is a good indicator of what should ultimately be "simple" to do and what is ok to be a bit more complex. In your case, that might rank out to be the following: 
1. User gets widgets/widget order
2. User changes widget order
3. User adds new widget
4. user removes widget
5. app is removed

Additionally, you want to limit the degree to which your system relies on consistency between these records. Don't put yourself in situations where consistency is difficult to write/maintain, but on the other end also don't build solutions that require consistency to read/function.

Lastly, be sure to add some sort of versioning aspect to your records so that you can apply condition checks when performing complex operations. This can be a version number or a `lastUpdatedAt` timestamp. If you maintain the `homescreen` record you'll need to get the list on the record, remove the reference to the app, and write it back. This is an unsafe operation, it could be operating on stale data, unless add a condition statement that checks that version/lastUpdatedAt you're updating equals the version/lastUpdatedAt you retrieved.

If we took the approach of maintaining order in a "homescreen" record, and also maintaining reference usage records, your database operations remain pretty simple. 

User gets widgets/widget order:
- Get homescreen record

User changes widget order:
- One change to homescreen (assumes the user is not able to "add" a widget in the same operation)

User adds widget:
- Check if app is "active"
- Write new app/user reference record
- Write homescreen record

User removes widget:
- Check if app is "active"
- Write new app/user reference record
- Write homescreen record

Remove app:
- Make app as "inactive"
- Crawl app usage records for usages
- Use identifiers on app usage records to also find associated homescreen records 
- remove usages and homescreen references
- Remove app from platform

**Reply by zirkelc** [reply]
Thank you for this thorough answer, that helped a lot!

I tried to translate your ideas and use cases into a ElectroDB model and implemented the operations. Could you take a look at this [Playground](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAYziEAwqlxUokgIJUq2AEaoqcACqbppAL7cMNmAARIhwRo6wTkEA3OJGUHxWYFQMTgikALzBmlCoGAD6KWlw+XwMknAxcQlJlphgWXxwLKrqWgAUomBKXd1gkrCIAFw9fX38GlSaI0F1GEH4vWP6mLhYCTMAjAtLfcQYhiYzKbgA1gzS0kG7PotjDA7OrngjirvdbmXqAJIwr+99KbeGZWbB8ADmOzGY3IaCw5D+yQw6Du0NuALmv3+0MBXjgIMcWAhUJxZDgcIRI0c6AB6JxqGkMAecBg9mxpKB+OCfFQkicmBJOJYDyMdBmIEF0NhqHhLKpyNMAO68C4DEoVBG7QAlA0AHxgAAizIAdHwoCxtaiccQNWBtXrDSazRatVa+n5oTc3UT4AAPF6jaFOTQAVX2WMDOOkp3ZpO4OAQiKC0clOPiMigazcIwA2kFPgwfjAggBdJVgOmks6x0lcBNJs6p6Hp2RZrl5zHFstxny0t297qV5RFdIV0Ra2LVRIKC7SBpgJotNSTTSdbpvMYDeAIGvdCZaGazpsrDBrDbBbZu7r7Q5c-MMM5Hr27B6ElxuXC75L8QtUCMbjk8QJMFISvPppVlRFqUVHFK3uS5-3LTlgKJUDywgykkRpWCwLKCovw+IDglBVCmzgvoGHCLAVgI5IiKCJwoCgdJCybZU4FVdURlVBBiDA8jugZJk3FZW0AJxZDuV5fl5jA7phSoUVxTYskKTlLCYNJFU1WoTUdQyfUjTcU1zUtcsbT0h0jLgEyXX4nFnzGH04H9T9Iz6YNbAQxFxOhaNaO6OtECTFM5L6FtM3Ydt8x-ItS3LAS9hjdycSCxMQVOFTugittcyCWdfnins+xuXoh2SVJR17CcpySOgbDwIxyH4edF1aFc13crdhhSsB9ymGZ6oqXAmpIPgmxvYw7xOc5LmPVZ1j4LZrjGcjXyeD8vwLItaMkoISOJMKMPU6D+LAztdvog60LjY6oIVM6ATYGBwTgGgAHkMHgDBaOAYBWCwF63t+XAkPohAsCsFSoskNzfIkq7CWJBLHvpRlmVEy7gSkvkBTChSlOCCUjvJGVMNO8ttK4u19MMp1TNdcy3ssgzHWM50zNghzVu9AQXIDeGwGDMNMEQuN-N6sYcqivLtr-LswsC+sZhTFHzOSwWpZsVsZbADtwwV8sleCjKVtJAc0X7MrUWHSrfHHSd4mnAGgYURpmnajpel87qd16-rpmCIbGua8awMmo5ghmo8wJPM8lovM3BzddasHfAWATlsXAOx-akZu0k7vlFESqFQHXvlrG72ulSi401HoQuyW6NzmuSbU+6S5wgECp8sHW-z2vScg4vNJ5zP2HSKuUMO0voTWAAvLlNZb6LcEkBgEAQBZggqGAsF5HeggQJhXtLMAHzAJ2rAbsYhIxtlm72nlcdk8sCbFImh470ewqp3SaZWXpnZJmtp7Ss2srZTmaJHJ9Gcq5L8wYADq5dgZ93FhrI2V9taRWzHrGKXx5alkVvGE2wRVY9jCtWZu2UcG5Xwc9CuhUSwkLSg2TKCU55jGFgbWi8CZjgjWAAJh3qScsEsV60IzPQvMWdDZxmNulYIgisBCIoebKhmCFHYOkbrfWot5EKLYQI4RjZOHdxxJ5byfC+a+hMVgS84itEKOlng2RsUiEsKwcY5RaxNjqIsVWZxcZXHRV7sQ7xytfEOLMcVQJg5rbewqsUMcNVRDXwUBHTIC4PZ6AOFNTqvk5jLgPGAOYKdLglIGhfS4bpg4jVDlUwO9TRr8DdIwt6JTA4dKoIk9cySqqLDSRk78hDfjzgIb+AAtNsWIIzOwTLmDMqoIze4TNnMsuZNQFA9PGVkIIPTNnpO2S7CuuBPrfXnHmQ5ABmI+hyRFllEKAMAXRwBzDAFAbwiQwAtNDq84QYg-pgCmfqCuvyGoNLGkLTQZSDbdFEH8savRjQV3aMODxvxCALJ8FqFF4IoDaliMC0FYBwU9KEMGOFBjRA9JRWgTAmgUU8IMei0Zv4sXUowOM3F+LCVpJeW8rlV9GAQjwKct6nyvqYDeUC-6pL75uAhcNVpfAJU0ERZC1VKLpAijoGyuR2L4W8u6MaG0bKKUXMwLmY0trLXSowIQXZMASwVjxaalgdBMBwHaGyxVLJ7AVkIMockbqHTknaP60ShAAAMWodQkv1HMAAElq0Oxoo32D5US8QgrhDCoYDAQQ6rZWiCycaRwhZcCUQ0AkFFLBsBuF9coWcTSnWoKoF0wgSL+BNLDazHMuxE0irCKcMAWAuA1LnJDC+VEVgvkqW0KYAJjSilHQazFMBCBrJNdCY0nrvXNtnRoFYQalB9WQP2-UEaa3UVMBpeNK70ySHYNA7ow7RrMnVXaWcwAPnkC4N6vgJh3V9B6V0p95BmQbrGVu9VnLOzbu8oQSYU9gheSuIQRe1cN5byCG6p9NgX1UDfWAYd-rlUhzGiMWc-AYDwZgLsHtfAmkrt1YpfVGLYNGoMQRnExpaMCAtR285DrczOtdbusYq6iOvsZt0EsoHUX8snHm95+wyQDBWEIOlgKy0GCmhWjAVbb11o9Y2n1bLW1Lu0OqrtlGoW9ps1evWQ75X6jXUYMdE6p3jqELe+d8FpCsf45504MGOVwZ3aB6TB7yBHoC74YNF6XM3rnfe6Cj7QuyZI-Jvow7yBadILpsDHaIOhag02rjkX22uwQwbJD0hOWobvBhne2GQS4e3nxvdz65NunI+jJVzGRiFagKe51TG01jRC3u9jooItFh49y+jUm+hmresJ12onvojHtd9Y0dZqCYHaIDB0Z2ACEGQsjOrADqNbpq+u5dRIp7NE4EW5vAEK-9cAiv+cuACkQzz3NEElbOZIUBx1lHS6IWcOq9WLflo1nlSnzUtvS9xTexAevKZzcDkFYLJUUu4HIKdvxjnO2Jzdjt9L0AYCZaaqxTWYCI85dFt7qn-qswpcaYSDBjQb2kL6nptN1UZoNni+ZBshBZH0RgZZJZoivPx9zkTvOHj88F8Ljtoueni4MZLk5zqZc5JaHoEj1yO0K-e7mkH4LmNCC4KTzsoMRkO-nMx-Fm2XcC4uL6zsovWdwZxfGpTBK8cjN2W4WGVzqtLYYxWV1ZGuf6h53zindVpv8G+NHk3OY49I65Zy-b9QfCK-fSnhzqrcDq5SOIctlbEimb4PWizCXF0rlqxXezzG+24odIOsYH6vVefHZO8HM6iSJYXcFmzT6R-hYL2z7yPXYtevi1Z9LZ6Q2Xv76zNLJ6MsKiy71nLZk3ME80+N4rIm3MAH4hZ6rANuN6pBWxrGKHf3Ytr90dpz79oQFdLXU7aPUXd+MrNoBnBRXnUIKrdlHacdaPCtTdLvNBEYGGX-OrVbGLUkGTSQYje0AFaERmC-BVIbUgB3UbX7a-b9HgQIEvDAe-R-DjMpcgsAd-NOdIL-aTW1Zjf-WGIAv3EA37MAuMArag09QsWFGGT5SdY3fwegkTK1RguMXvOfbRY0ebTjeAyuRA37ZA7jIvREDAkPLBDbEjZQBg9ApAhgw7LAY7DAU7ejVXLAmvRgXAJw+NVfXAp7bUAEEgofEHMbSQy4GfWbaTYIn1JfKLFfB7MAPAgggIsAV7U1cPd7PoIAA)?

There are a few things I'm still not sure yet. I have used transaction write if a use case spans over multiple entities (user adds or removes widget; user removes app). Especially the last use case (user removes app) is a bit complicated.  It seems transactions don't support batch delete and batch write, so I `map()` over `widgets` and `homescreens` and delete/update these in single operations which I spread into the transaction. Not sure if that's a good idea though.

I added a conditional check  on `homescreen.updatedAt` for the use case _user changes widget order_, because I `get()` the `homescreen` and `update()` it on two decoupled operations. But, do I also need to check for stale-ness during the transactions?

**Comment by tywalch** [comment, marked as answer]
I threw this together this morning, I only used the playground so don't consider it tested or "complete" but it should get the point across. You want to ensure that your operations remain idempotent so the caller can retry when failures occur:

https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAYziEAwqlxUokgIJUq2AEaoqcACqbpp1eq0BJN0lzcgY3NQ0qTUC4YN9IzQAlOCM5GDAAX24MGzAAIkQUx1gnPIBucVS+KzBUVCx0gF4wAAoASjBGgD4wAHIGJyNeitEqmqoBhFJmvM0oVAwAfQmnKcW+Bkk4csqoaoVLTE6wPjgWeK0W0TAla5uwSVhEAC5b+-v+BNe8w4w8-Du730mFwWD23wAjP9AfdiBhDCZvhNcABrBjSaR5GEZAHvMKOLAuNy4V6KGE3NwbdT+GCk8n3KLeb5WbB8ADm0Pe73IaCw5FpYEc6FxXJx9N+NLpXIZXjgzIJ7M50rIcF5-NeQrg9LF0tQ0hgYTgMHsUuVjLl+T4qEkTkwSulLDCRjo3xA9q5PPq6sFGGF9Ju8C4DEoVFe7U6PQAIoaAHR8KAsdoi6XEUOtDrdMDRtxxhNJ7XS7HJsBYARwAAeeFN7ycmgAqnDJW9ldIUdXlVwcAgBXlW+7pakZFBQW5XgBtPKUhjUmB5AC6-syxfeqPb0s7iB7qP7XMHshHFonEtnC+VmW1xYydx1SkFk1IV7aI1Gexq6Okx1O5wilzuZPejzwAga5gJ8WjfO+O7AhgoLgvkULLkQBjGBak4MKikFFjC+LOK4VbNu8U4ziBFKyvKrIcohNyenyRoar6pgXvS75Nv+ZpkfkLKlpRi40d6mpMdKGxbCRgocXkXGKohN54kYGjAqJ5rfE4UBQFM047gGcBBiGrxBggxDSYheoGm4xppmx0pKZa1q2n8VGsE6Lr5G6Dl8XRPp+megbBtQYYZlGsbxombQOam-kRlmQV5qFBZclh7ylvAlYkgR9y1rYGKsYuraiTcG7dt8fYOTce7Duwh6Tvw05UDS86LjJXKrml65dluKKafcZUHuOeQsSeDWCfcV43DeygrFM55PuI5pgAA6g0bJwAozRsWwMBLbVAqSWyFQ3P1rw7XtgrsFMh0KrtdyggAXqhuCSAwCAIHkYAAD75FsMBYNaL3vXkCBMEtOxXi++xgHQNh4EY5D8J+ZwXFEVw3JZDxPMBLWgX4UTfBDWy4NDJB8DucIIqhyJohiUEgmCfCQli7yNThhJ4alKMUtVxEY6RTKcRdnXuQKmpGeKjYCmzMo8xJfNuaqXoeULQ3vOtm2s4u1kWFYNj2ASRIeLKAA8C0bctY5zl0LR9Xwmh5LFZ4C-RXmisZ+qGuZiniVaNp2g5jpUM6rr87LtGCwxDk+bp6aRdmcC5iFYXLRFmbR7H+bSlk8UM8WSUVvh4u1g2mDZWeuVc11Nj7hVvVEVt84ldwbVFR1DVhW2peleX5WjmAR6i7Xi75Q3nFN2eI1O9i14iuN95TSMYwKO+DYMEtcPfljmhI2lgEvBjYHY-kC+4Ev2yISTKFIuhFOYoh0GwbT8H08NxZM7rqtWRzW3u5LO2B2q8uh4r9xlbLSLuxL+0teJB34v-NOxlRafzuuAu2kC-6O0zsxLKYs1biW-jLX+IdUGP3FC7MyJpS7WTyJ7OynVfb+xcj-OW+DGLeW0r5NM4Yk7RTjoucKkcOE5mCqnUUCV7jZxSiBWsRtNogOlCXcW7xuqV27lVKkNdCA-F7nOOuBUex9mbtw1uciy5Dh6ko-q-x8hAJrpo-u9dNzMmHsqUeaDpT5zgaXUR3w2SggAEzmOVDlAxNiFFdwnNXOq1izwDzsfkLxWBvG6JHi3PKYBgmVWPH3SJtjCoxJ8duQaMD6QZQwaJDxOSsAIQCck1JVd37hK0YPPIsSIQJMcUktuKSO4mInGYiJkTtGeNBBCPJI8AGjzGneVYD5RDTVBjUU+JgV5gD0PCFCG8Ua-CfhiYsuMoYwz4Js6Qi8loinGRNB8AIZmzQABKQ3xnsyRy1jg-iiMkVIGAYD63NFALg4NbkE34F0McvRLG4F6HOMcAAGOcIxZoqHLNIac8AYA3Lxv8vgDyVpgDuAAMjACi3ZhMMU4oRtEIInzZTfLAO+LoIxQBYuELUOEYAoDeGqL81Fez6UiFEMAYAYAAC0PRNrsoJbDWsjLC7pCxbNAA4stfFdzCYAHlpAaFfMcNiYTtoXWOsec6FEKgg3QpoPgRhuCoFNWqvgYBNoKrRRi3ALQWVWtSnKqgdq9kqpdW0V4AAFHIkgsDEH1nChFZZkV-PuYtE2Ztmxz1vFqwgx5MjHGdTTXAx143KFMgwFNzQGCOnYEhFZJgYy7xwLgGMOzFX8BhDGTaLRxq1JgEm0WmQ2h1rZFAdoIx25gxBa8UNiKjQesJdGqgptjim2OncLAPyWgAEIc0dBRuQKgCxrUguOiDfahaFD+psEGmOj0EAtBzTGEFMYHrSBaMa01rRLEBTSlmqlGC82sHHcdPtNRs1hFzVkfNe7i2kzLVjCtMZ3z1uWo2190gaSEC1e2+t3bprvBhLywUdBSCIGICk9CpAg1UqIFgNk1rcCQw4NxZl1qOCkCMOpa1epiPwHBuhG1exSAsByOydDfKsAqkHFsAQYAEBQDZGyKjiK7woio+QVICxcMwjna0HNYBsW4vPQwOSWBgQrvpJe6Qlg6CNsXDGMzliHJmZjDm6SHaGbXjaDM6iy0N0fuNjQQ14g6XXHAL8FJjB2R4Dc8KtImAfNiAw4KpCChq1ouC480LGBRCzT0O6yNyr3mYC9emjVdwtX6u4rqtxR07gDrxel-gGLp13ABlYOsxCjT2AAPyvEoZgTzPK+UPRRKQFlmAwg0xLPAWQlIFAsCw9a-coJJmiDvWargFrtN7Gi6O-gTrVXpteKl1bfAlWZYwNl18env0KCbSo+DEqMAXZBYQWrVB6umUawoADzKNuvkza+U7YBVMvYLQwItsW9l1r1HCKgMHE2XZpO2utqYYMguh+8C9WHyAtBgyZV29gly3njFQOFQaaC3bcO2yK9I7sPYxytRozQLWBlLEaRcTWThQFx+WfHjr0ckKoHZ5UrxqBwBaBzp7t30L3Ya+ZbnNwJdgGQ428guBZDVAtP0J66wzi9AyKh+lKp10YGtTmjr3mGV+YYDAQQ8WqDheS7KMAthTc7YxYdtlq07h6qIDqvLzaCvsmOmTsXzXWu2Xa3cCs3g5K2G08CFrYAVJqTgNOY6B8j6vCVYGqg+t7fjsIMC8dNJeg0tEEa3AJr5uLatVSu3FX0XjvWy614tuI0crHe5x3uBjsdLBsoRPxzBTNtbZK4XdW-dUEICHoo4f5LTE8g+VNb3qgfY77Bi7N0fCREmr9jERy4Dz5qJYqH1P6gwB7XcNdrn5kx0cNOQ+S2+B1q4xVVHnetmwc34QQHhNieZjHPSSLPRnQpBRCWD8u+CWEIFphPugtIKZn-kYCiODr3ovukBrqZuNpgPzjBmATpucreKqB-j0KqLehHj4KPmHoQY5lATYKnoIu8D-ikqEETpYq0O+MAH5uQFwKgaanAFLvtBvofEtFAXQfzmdjVNdjni2pDmIf1AjsqDGIJuwFQfcDQYLiKjWnfO+PwOkLvjAPSG-rWmeDGCDpgGDkITOH3ldogVwfcBBhiOoXDuOqlGOI-nBmIcvghqdD4JoZkHOFIdKEjqgQ-rUEPljsoDjnjlYLQHzrgVrmeL7o9uZJ0FTrUGWJ2KcFoZkjcIziEazmEQLkPhYdKLzm4DkbEfYAPqLsUVzouHkdLrIWDrbPcHOFLjLtNCDIbr5kyuQI8MCEIJYpbrNMkJ0XABns3rPkIM7jcK7iVuzOdtqgaqVqIV7pdNwU4QsT7iLuTpzgHl7BgMdMQVQOPpgVHjHgxh1nNuapaoNh0VAMCEMQ2mmq+K8P0VcYMZXg7iMW3i+poQhvAceIQP1KUesULqBPCmPoQccJqO+ncXPncPGjodaoBv9goGfqBpEOBrCVBkYRDseBrk0cdMpoumiapq9O9ASX+voYET0DERTm3hSHQDkOcF+KoBgDkBgC0L0P4HwPoI9A0MoWir0Jrjuu3jvnYccCSRMBenYTGJ2HziyQwZmJYuKe5lDgugkZoU5trqfshKWhftUBgXsLftgIUegU-l3j4LCZEV-tKDQdAQAcpsAYRhgcCBAVAVhjAXAdMb8W+kgXoSgSjugaCTiNgcgJEfgQ6UQcCSQRPmQXoTUTBnLgrsQK8MrggIsGpDAL0O2ohFaQIebowRiMwe0dpOwSYBYSaaZkBMtIIT3u6ebhdn8eId4VyDIRQXIcoHGa+Erieimd2OmRrpmXylFkobCa8GocJpodoZXqZgYfIG6cIWIT8ebkIF6dIT6WgcoILnEQGcoBERrpFHzkURTv8bkVUU2ZIJQa2XgPGR2SrqmT2XUTcA0Z2ihp5q0ZdiqAMaARiFymIH0bEE8ZlNIC3rlksU2JMVWbOSsQXjNNbo8cCP+WYNOCYFMDACqv1mXs0LYHukaPrMkDrnwJ4N4OSt4JSpcdcS8eOl0ICr0DmmCiMKcQtucctiRXAP+TXptmADBcxRiC3j6mAAeoGsGhxXBQhYUMhd4BgANnsKbD0GxC+nWYhi9pCRmncHcH9kWkieWngFYZAYjgin7MZsYVtB6U4Q2TGLDp3oQXpI9LhkuTcDieICdguccKpYiZqTHBpbgHWiaTGGgJgJoHWkUk4TOSYQgSZV2jBgiktKlEmbeb2oKQoEYMJUhSheJbXuxb+bBRiPBRwUlWJRJXwJOs0JOsdFwHIK0PGgwZSpesus+p9reHJd8UPqYSIYqYgccJYl+lEWABhtYDAFAK8GMA0EcFxu+FRpRkIAlU9CWNamuhgJoCJmJsYDCC+gldlUaIQD9k5UBkxTcdBuLHWfSImiLP3vpqIYhJSRsQEeUYhLsfsQpNwFZUwsNGqTcHiStYhUaNSV1IlUaMlXlZWoZrgMZsuh1WMvZYAWVd9aJahemjGFMOyBwJ9SfrrnhqtVDSlemtumDc5cBihMiewOBu+HWuWYUQZbWZ6Y0WFWqUjdatVhkEAA

**Reply by zirkelc** [reply]
That's awesome! 👍  Thank you so much for taking the time to help me out!
